packages/pg/lib/stream.js

```diff
@@ -0,0 +1,176 @@
+  constructor(ssl) {
+    super()
+
+    this.ssl = ssl
+
+    this.writable = false
+    this.destroyed = false
+
+    this._upgrading = false
+    this._cfSocket = null
+    this._cfWriter = null
+    this._cfReader = null
+  }
+
+  setNoDelay() {
+    return this
+  }
+  setKeepAlive() {
+    return this
+  }
+  ref() {
+    return this
+  }
+  unref() {
+    return this
+  }
+
+  async connect(port, host, connectListener) {
+    try {
+      log('connecting')
+      if (connectListener) this.once('connect', connectListener)
+
+      const options = this.ssl ? { secureTransport: 'allow' } : {}
+      const { connect } = await import('cloudflare:sockets')
+      this._cfSocket = connect(`${host}:${port}`, options)
+      this._cfWriter = this._cfSocket.writable.getWriter()
+
+      this._cfSocket.closed.then(() => {
+        if (!this._upgrading) {
+          log('CF socket closed')
+          this._cfSocket = null
+          this.emit('close')
+        } else {
+          this._upgrading = false
+        }
+      })
+
+      this._cfReader = this._cfSocket.readable.getReader()
+      if (this.ssl) {
+        this._listenOnce().catch((e) => this.emit('error', e))
+      } else {
+        this._listen().catch((e) => this.emit('error', e))
+      }
+
+      await this._cfWriter.ready
+      log('socket ready')
+      this.writable = true
+      this.emit('connect')
+
+      return this
+    } catch (e) {
+      this.emit('error', e)
+    }
+  }
+
+  async _listen() {
+    while (true) {
+      log('awaiting receive from CF socket')
+      const { done, value } = await this._cfReader.read()
+      log('CF socket received:', done, value)
+      if (done) {
+        log('done')
+        break
+      }
+      this.emit('data', Buffer.from(value))
+    }
+  }
+
+  async _listenOnce() {
+    log('awaiting first receive from CF socket')
+    const { done, value } = await this._cfReader.read()
+    log('First CF socket received:', done, value)
+    this.emit('data', Buffer.from(value))
+  }
+
+  write(data, encoding = 'utf8', callback = () => {}) {
+    if (data.length === 0) return callback()
+    if (typeof data === 'string') data = Buffer.from(data, encoding)
+
+    log('sending data direct:', data)
+    this._cfWriter.write(data).then(
+      () => {
+        log('data sent')
+        callback()
+      },
+      (err) => {
+        log('send error', err)
+        callback(err)
+      }
+    )
+    return true
+  }
+
+  end(data = Buffer.alloc(0), encoding = 'utf8', callback) {
+    log('ending CF socket')
+    this.write(data, encoding, (err) => {
+      this._cfSocket.close()
+      if (callback) callback(err)
+    })
+    return this
+  }
+
+  destroy(reason) {
+    log('destroying CF socket', reason)
+    this.destroyed = true
+    return this.end()
+  }
+
+  startTls(options) {
+    this._cfWriter.releaseLock()
+    this._cfReader.releaseLock()
+    this._upgrading = true
+    this._cfSocket = this._cfSocket.startTls(options)
```

Comment:
This will need to `this._cfSocket.closed` then handler to be reapplied to the replacement socket correct?

Reply:
Good spot!